### PR TITLE
Form migrations: Dashboard- and TimeZonePicker

### LINKED ISSUE #22404: Grafana UI: Enable using new form styles in Select-based components
**What would you like to be added**:
Add `useNewForms` prop to `DashboardPicker` and `TimeZonePicker` in order to enable using new form styles. 
**Why is this needed**:
In order to introduce new form styles without creating new components, a temporary solution is to add `useNewForms` prop to some components to enable toggling between old and new styles. This is particularly important for components based on `Select`, as there are a few places where old styles still need to be used and we don't want to create separate new components that would use new `Forms.Select`.

### LINKED ISSUE #22410: Grafana UI: Forms.Select doesn't work properly with option group
**What happened**:
When passing `options` prop as an option group (format `[{label: string, options: SelectableValue[]}]` and a `value`, `Forms.Select` component fails to set that value as selected. 

Seems like the [custom code](https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/components/Forms/Select/SelectBase.tsx#L201) for handling selected value doesn't account for the option group format.

**What you expected to happen**:
`Forms.Select` should set its value to the `value` prop when option group is supplied via `options` prop. 

**How to reproduce it (as minimally and precisely as possible)**:
Can be reproduced by using `Forms.Select` instead of `Select` in the [TimeZonePicker](https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/components/TimePicker/TimeZonePicker.tsx).

**Anything else we need to know?**:
The value is selected properly when custom `selectedValue` is bypassed and `value,` vs `value: isMulti ? selectedValue : selectedValue[0],` is used [here](https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/components/TimePicker/TimeZonePicker.tsx).

### PR DESCRIPTION
**Pull request by Clarity-89**
**What this PR does / why we need it**:
- Apply new Form styles to DashboardPicker and TimezonePicker
- Fix Select to properly handle selected value if `options` is an option group array (for single Select)
- Move type declarations from SelectBase to a separate `types.ts` file to make `SelectBase` less cluttered and avoid circular imports. 

**Which issue(s) this PR fixes**:

Fixes #22410 
Fixes #22404 

**Screenshots**:
Screenshot 2020-02-27 at 8 26 46
Screenshot 2020-02-27 at 8 26 41

### CODE REVIEW

File: packages/grafana-ui/src/components/TimePicker/TimeZonePicker.tsx [hunk 1]
```diff
@@ -31,11 +32,12 @@ export const TimeZonePicker: FC<Props> = ({ onChange, value, width }) => {
   });
 
   return (
-    <Select
+    <Forms.Select
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/TimePicker/TimeZonePicker.tsx]: Using `Cascader` here instead would make more sense

File: packages/grafana-ui/src/components/Forms/Select/utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+const options: SelectableValue[] = [
+  { label: 'Option 1', value: 1 },
+  { label: ' Option 2', value: 'test2' },
+  { label: 'Option 3', value: 3 },
+  { label: 'Option 4', value: 4 },
+  { label: 'Option 5', value: 'test5' },
+  { label: 'Option 6', value: 6 },
+];
+
+describe('Forms.Select utils', () => {
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Forms/Select/utils.test.ts]: Awesome with tests and the utils!

**Review by tskarhed — CHANGES_REQUESTED** [review]
Good work! Thanks for cleaning up Select a bit, it's really needed! In regards to the `TimeZonePicker`, I think it makes more sense to use the `Cascader`. That probably applies to most Selects that need grouping #22410. Unless we have non-expandable grouping, I really can't see many cases where `Select` would be preferred over `Cascader`.

I also had this result in Storybook
Screenshot from 2020-03-02 14-36-57

**Comment by Clarity-89** [comment]
Thanks for the review! Didn't know about Cascader, will look into that.

**Comment by Clarity-89** [comment]
@tskarhed I've looked into Cascader for the TimeZone picker and this is how it'd look. Not 100% sure it's the best solution (particularly the scrollbars seem a bit weird), but I guess we could give it a try.

Screenshot 2020-03-02 at 18 29 59

**Comment by torkelo** [comment]
I agree this should use the cascader.

**Comment by tskarhed** [comment]
> @tskarhed I've looked into Cascader for the TimeZone picker and this is how it'd look. Not 100% sure it's the best solution (particularly the scrollbars seem a bit weird), but I guess we could give it a try.
> 
> Screenshot 2020-03-02 at 18 29 59

The native scrollbar only shows in Storybook, it's custom in Grafana. We should probably do something about that.

File: public/app/core/components/Select/DashboardPicker.tsx [hunk 3]
```diff
@@ -46,25 +47,21 @@ export class DashboardPicker extends PureComponent<Props, State> {
...
-            loadOptions={this.debouncedSearch}
-            onChange={onSelected}
-            placeholder="Select dashboard"
-            noOptionsMessage={() => 'No dashboards found'}
-            value={currentDashboardId}
-          />
-        </div>
-      </div>
+      <Forms.AsyncSelect
+        size={size}
```
Comment (tskarhed) [inline public/app/core/components/Select/DashboardPicker.tsx]:
```suggestion
        size={size || 'md'}
```
Probably needs a default fallback, missed this the first time
Screenshot from 2020-03-03 14-25-10

**Review by tskarhed — APPROVED** [review]
Just the comment about the fallback, missed it the first time, then you're good to go. Approving so you'll be able to merge when you've fixed it. The rest of the ReportPage should be migrated to the new form styles too ASAP. I'll have a look at it.

File: public/app/core/components/Select/DashboardPicker.tsx (hunk 3 above)
Comment (Clarity-89) [reply public/app/core/components/Select/DashboardPicker.tsx]: @tskarhed  Yeah, good point, altho I think these fallbacks are better handled via `defaultProps`/default params, so they are all in one place :) 

I already have a PR that migrates the rest of the ReportPage to the new form elements, just needed this one to get merged first.

Comment (tskarhed) [reply public/app/core/components/Select/DashboardPicker.tsx]: Good point, that's awesome too! Can you add the default prop in this PR?

Comment (Clarity-89) [reply public/app/core/components/Select/DashboardPicker.tsx]: Done :)

**Pull request merged**